openmdao/components/akima_spline_comp.py

```diff
@@ -0,0 +1,460 @@
+"""
+Simple interpolant based on Andrew Ning's implementation of Akima splines.
+
+Includes derivatives wrt training points.  Akima spline is regenerated during each compute.
+
+https://github.com/andrewning/akima
+"""
+from six.moves import range
+from six import string_types
+
+import numpy as np
+
+from openmdao.core.explicitcomponent import ExplicitComponent
+
+
+def abs_smooth_dv(x, x_deriv, delta_x):
+    """
+    Compute the absolute value in a smooth differentiable manner.
+
+    The valley is rounded off using a quadratic function.
+
+    Parameters
+    ----------
+    x : float
+        Quantity value
+    x_deriv : float
+        Derivative value
+    delta_x : float
+        Half width of the rounded section.
+
+    Returns
+    -------
+    float
+        Smooth absolute value of the quantity.
+    float
+        Smooth absolute value of the derivative.
+    """
+    if x >= delta_x:
+        y_deriv = x_deriv
+        y = x
+
+    elif x <= -delta_x:
+        y_deriv = -x_deriv
+        y = -x
+
+    else:
+        y_deriv = 2.0 * x * x_deriv / (2.0 * delta_x)
+        y = x**2 / (2.0 * delta_x) + delta_x / 2.0
+
+    return y, y_deriv
+
+
+class AkimaSplineComp(ExplicitComponent):
+    """
+    Interpolant component based on Andrew Ning's implementation of Akima splines.
+
+    Assumes a uniform distribution of points. Control points fully span the distribution.
+    Output (computational) points can be at either the end points or segment center
+    points.
+
+    Attributes
+    ----------
+    x_grid : None or ndarray
+        Cached training grid.
+    x_cp_grid : None or ndarray
+        Cached interpolation grid.
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Initialize all attributes.
+
+        Parameters
+        ----------
+        **kwargs : dict
+            options dictionary.
+        """
+        super(AkimaSplineComp, self).__init__(**kwargs)
+
+        self.x_grid = None
+        self.x_cp_grid = None
+
+    def initialize(self):
+        """
+        Declare options.
+        """
+        self.options.declare('num_control_points', types=int, default=10,
+                             desc="Number of control points.")
+        self.options.declare('num_points', types=int, default=20,
+                             desc="Number of interpolated points.")
+        self.options.declare('vec_size', types=int, default=1,
+                             desc='Number of independent rows to interpolate.')
+        self.options.declare('name', types=str, default=None, allow_none=True,
+                             desc="Name to use for the interpolated variable.")
```

Comment:
can 'name' really be None?  It looks like it's used to create the input and output vars...

Reply:
Fixed.  Fixed all but question about what to call the component.